### PR TITLE
SSB Transmitter

### PR DESCRIPTION
**Pull request by corvus-ch**
Implements #117.

### CODE REVIEW

File: examples/ssb-transmitter/ssb-tx.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        WavSpec {
+            channels: 2,
+            sample_rate: file_rate,
+            bits_per_sample: 16,
+            sample_format: SampleFormat::Int,
+        },
+    );
+
+    // TODO Make this work with `ssb-receiver`.
+    let dat = FileSink::<Complex<f32>>::new(format!("{}.dat", cli.output));
```
Comment (corvus-ch) [inline examples/ssb-transmitter/ssb-tx.rs:128]: I was unable to decode this using the existing `ssb-receiver` example. Any ideas why that does not work?

File: src/blocks/mod.rs [hunk 2]
```diff
@@ -24,6 +24,7 @@
 //! | Block | Usage | WebAssembly? |
 //! |---|---|---|
 //! | [ConsoleSink] | Log stream data with [log::info!]. | ✅ |
+//! | [Delay] | Delays samples. | ✅ |
```
Comment (corvus-ch) [inline src/blocks/mod.rs:27]: Would it make sense to also use this delay in the `zigbee` example?

**Comment by corvus-ch** [comment]
Shall I create separate PRs for moving the `Deplay` block and introducing the `hilbert` filter design function?

**Comment by bastibl** [comment]
Hi, very good PR. Thanks! I have only two minor comments:

- I think, it makes sense to have SSB RX/TX in one example. Maybe also with a loopback flowgraph.
- I think, the Hilbert code could be more concise with iterators. For example, using [`skip_by(2)`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.step_by).

@loic-fejoz Could you have a look at the DSP side of things and why it doesn't work w/ the receiver?

**Comment by corvus-ch** [comment]
> I think, it makes sense to have SSB RX/TX in one example.

Done.

> I think, the Hilbert code could be more concise with iterators. For example, using [skip_by(2)](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.step_by).

Also done, but I might miss some opportunities.

To finish this up, I would like to resolve the incompatibilities between receive and transmit. So, if anybody can give me hint?

**Comment by loic-fejoz** [comment]
Thanks @corvus-ch, I have tested your code in this branch: https://github.com/loic-fejoz/FutureSDR/tree/test/ssb/examples/ssb

I had to made few adjustment in your code for it to compile (like f64 and f32 mixing), and also clarify the output filenames.

On the receive side, I had to make the `file_level_adjustement` a command line parameter so as to be able to hear something.
But overall it seems to work.

In my test the sound was not that great but I guess it is due to some saturation in the whole chain...

You can also test your result with GNURadio for instance:
image

**Review by ellenhp — COMMENTED** [review]
I tested this and had to make the same modifications, but I wasn't able to get anything resembling the original audio back out of it. Still not sure what's wrong though, could be on my end. I also noticed that the polyphase resampling stalls the entire pipeline for some audio/file sample rate combinations and it seems to have something to do with the GCD of the two numbers being small relative to their magnitudes. Does polyphase resampling misbehave for ratios with large integer values in the numerator and denominator?

For audio signals at least it might make sense to add a block to perform resampling with hermite interpolation or something similar, since I think that's a fairly cheap operation compared to polyphase resampling.

edit: I stepped away from my computer and realized that the pipeline stall may have been caused by how short an audio clip I was using. The FIR filter could be waiting for more samples before it emits anything? Would be nice to have something like #114 to know for sure though :)

File: examples/ssb/README.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
+SSB
+===
+
+Examples of an SSB modulator (`receive`) and demodulator (`transmit`).
```
[the remark this answers is not in the document]
Comment (loic-fejoz) [reply examples/ssb/README.md:4]: I double that.
transmit => modulator
receiver => demodulator
https://en.wikipedia.org/wiki/Modulation

**Comment by loic-fejoz** [comment]
@ellenhp  and @corvus-ch Here is a recording of what I can get out of it without spending too much time on it:

https://youtu.be/X4LkNk7YO4E

I am pretty sure it is not terrible due to some saturation. One should spend time on it, but on principle, it is closed IMHO.

File: examples/ssb/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (corvus-ch) [reply examples/ssb/README.md]: Totally a mix-up on my side. Fixed.

**Comment by corvus-ch** [comment]
I got it working. Thanks @loic-fejoz for pointing out the file level issue. The reason it did not work for you was a mix-up of LSB and USB.

I am left with one thing that puzzles me. The audio of `ssb_lsb_256k_complex2.dat` says: "… on 50.3 MHz". Yet the center frequency of `receive` is 51.5 kHz and I had to set the transmit frequency to 53 kHz. I also did a quick conversion of `ssb_lsb_256k_complex2.dat` to a wave file I could open with SDR++. There the signal showed up at 53 kHz.

* Is the announcement supposed to be correct?
* If yes, why does it seem to be off by 2700 Hz plus a whole magnitude?
* Would it make sense to switch `receive` to the "carrier" frequency as it is the case with `transmit`? This would remove a source of confusion.

**Comment by bastibl** [comment]
I guess the signal was sent on 50.3 MHz, but before it was digitized, it was down-converted. Otherwise, the 256kHz sampled signal could not represent the required bandwidth. In the sampled version of the signal, it is centered at an offset, that depends on the settings of the down-converter.

I'd assume the center frequency of an LSB/USB signal is the frequency of the suppressed carrier. So the different between RX and TX suggest there are some inconsistencies here.

**Comment by loic-fejoz** [comment]
Indeed @bastibl , usually we spell the frequency of the suppressed carrier for single-side band. Also interestingly @corvus-ch  some people take 2700Hz as the maximum frequency people can hear, and that the number they use in filter and such (sometime rounded at 3kHz). Have a look at pictures from https://www.sigidwiki.com/wiki/Single_Sideband_Voice  

I will try to do my own recording for testing purpose ...and I should really work on a SigMF block also!

**Comment by corvus-ch** [comment]
@bastibl @loic-fejoz How do we progress from here? Is there anything holding you back from merging this? If yes, is there anything you expect me to do?

**Comment by bastibl** [comment]
Sorry, I didn't know that this was good to go. Your last comment suggested that there are inconsistencies between rx and tx. Thank you!

**Pull request merged**